### PR TITLE
Add search indexes

### PR DESCRIPTION
**Pull request by tseaver**
~~Uses #1164 as a base.~~

### CODE REVIEW

**Comment by dhermes** [comment]
Ping me here when #1164 is in (or at least once that review is done)

**Comment by tseaver** [comment]
@dhermes Rebased after merge of #1164.

File: gcloud/search/client.py [hunk 1]
```diff
@@ -17,6 +17,7 @@
 
 from gcloud.client import JSONClient
 from gcloud.search.connection import Connection
+from gcloud.search.index import Index
```
Comment (dhermes) [inline gcloud/search/client.py]: Does `Index` belong in `__init__.py` as well?

File: gcloud/search/index.py [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+        :type client: :class:`gcloud.dns.client.Client`
+        :param client: Client which holds credentials and project
+                       configuration for the index.
+
+        :rtype: :class:`gcloud.dns.index.Index`
+        :returns: Index parsed from ``resource``.
+        """
+        name = resource.get('indexId')
+        if name is None is None:
```
Comment (dhermes) [inline gcloud/search/index.py]: `is None is None`

File: gcloud/search/index.py [hunk 3]
```diff
@@ -0,0 +1,150 @@
...
+            raise KeyError(
+                'Resource lacks required identity information: ["indexId"]')
```
Comment (dhermes) [inline gcloud/search/index.py]: I thought you were against custom `KeyError`s?

File: gcloud/search/index.py [hunk 4]
```diff
@@ -0,0 +1,150 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+"""Define API Indexes."""
+
+
+class Index(object):
```
Comment (dhermes) [inline gcloud/search/index.py]: This is just a data container for now, right? Can't be used to do anything?

**Comment by dhermes** [comment]
Sorry for the delayed review. I accidentally knocked this PR out of my task queue.

File: gcloud/search/index.py (hunk 3 above)
Comment (tseaver) [inline gcloud/search/index.py]: This is exactly the same thing we do for the DNS / PubSub `from_api_repr` methods.

File: gcloud/search/index.py (hunk 4 above)
Comment (tseaver) [inline gcloud/search/index.py]: The relevant index-only APIs are `list` (return info about a project's indexes) and `search` (return documents from a project matching a query):  I plan to implement `search` last.

@jgeewax seemed to indicate that the API might grow support for further index methods (there should at least be a `get` method, for instance) before GA.

File: gcloud/search/index.py (hunk 2 above)
Comment (tseaver) [inline gcloud/search/index.py]: 3b8b31c

File: gcloud/search/index.py (hunk 4 above)
Comment (dhermes) [inline gcloud/search/index.py]: How does one create/update/delete an index?

**Comment by dhermes** [comment]
LGTM. (Would still like to know about CRUD ops on indexes)

File: gcloud/search/index.py (hunk 4 above)
Comment (tseaver) [inline gcloud/search/index.py]: One doesn't:  one adds documents via URLs which imply the index.

FWIW, I think the API designers took their fingers off the number, there.

Comment (dhermes) [inline gcloud/search/index.py]: :+1:

**Pull request merged**